src/algorithms/calorimetry/ImagingTopoCluster.cc

```diff
@@ -190,21 +226,67 @@ bool ImagingTopoCluster::is_neighbour(const edm4eic::CalorimeterHit& h1,
   // layer check
   int ldiff = std::abs(h1.getLayer() - h2.getLayer());
   // same layer, check local positions
+  // if (ldiff == 0) {
+  //   return (std::abs(h1.getLocal().x - h2.getLocal().x) <= localDistXY[0]) &&
+  //          (std::abs(h1.getLocal().y - h2.getLocal().y) <= localDistXY[1]);
+  // }
   if (ldiff == 0) {
-    return (std::abs(h1.getLocal().x - h2.getLocal().x) <= localDistXY[0]) &&
-           (std::abs(h1.getLocal().y - h2.getLocal().y) <= localDistXY[1]);
+    switch (m_cfg.sameLayerMode) {
+    case ImagingTopoClusterConfig::ELayerMode::xy:
+      return (std::abs(h1.getLocal().x - h2.getLocal().x) <= localDistXY[0]) &&
+             (std::abs(h1.getLocal().y - h2.getLocal().y) <= localDistXY[1]);
+
+    case ImagingTopoClusterConfig::ELayerMode::etaphi:
+      return (std::abs(edm4hep::utils::eta(h1.getPosition()) -
+                       edm4hep::utils::eta(h2.getPosition())) <= layerDistEtaPhi[0]) &&
+             (std::abs(edm4hep::utils::angleAzimuthal(h1.getPosition()) -
+                       edm4hep::utils::angleAzimuthal(h2.getPosition())) <= layerDistEtaPhi[1]);
+
+    case ImagingTopoClusterConfig::ELayerMode::phiz: {
+      // use average phi to calculate transverse coordinate in rotated frame
+      auto phi  = 0.5 * (edm4hep::utils::angleAzimuthal(h1.getPosition()) +
+                        edm4hep::utils::angleAzimuthal(h2.getPosition()));
+      auto h1_t = (h1.getPosition().x * sin(phi)) - (h1.getPosition().y * cos(phi));
+      auto h2_t = (h2.getPosition().x * sin(phi)) - (h2.getPosition().y * cos(phi));
+      auto h1_z = h1.getPosition().z;
+      auto h2_z = h2.getPosition().z;
+
+      return (std::abs(h1_t - h2_t) <= localDistXY[0]) && (std::abs(h1_z - h2_z) <= localDistXY[1]);
+    }
+
+    default:
+      error("Unknown layer mode for same-layer clustering.");
+      return false;
+    }
   } else if (ldiff <= m_cfg.neighbourLayersRange) {
-    switch (m_cfg.layerMode) {
+    switch (m_cfg.diffLayerMode) {
     case eicrecon::ImagingTopoClusterConfig::ELayerMode::etaphi:
       return (std::abs(edm4hep::utils::eta(h1.getPosition()) -
                        edm4hep::utils::eta(h2.getPosition())) <= layerDistEtaPhi[0]) &&
              (std::abs(edm4hep::utils::angleAzimuthal(h1.getPosition()) -
                        edm4hep::utils::angleAzimuthal(h2.getPosition())) <= layerDistEtaPhi[1]);
+
     case eicrecon::ImagingTopoClusterConfig::ELayerMode::xy:
       return (std::abs(h1.getPosition().x - h2.getPosition().x) <= layerDistXY[0]) &&
              (std::abs(h1.getPosition().y - h2.getPosition().y) <= layerDistXY[1]);
```

Comment:
Who uses this? This is a dangerous mode since it is global XY positions instead of local XY positions (as the same-layer mode of the same name), and it will therefore only work for endcap detectors.